File: Java/app/src/main/java/net/consentmanager/kmm/cmpsdkdemoapp/MainActivity.java
package net.consentmanager.kmm.cmpsdkdemoapp;

import android.content.Intent;
import android.graphics.Color;
import android.os.Bundle;
import android.util.Log;

import androidx.activity.ComponentActivity;
import androidx.annotation.NonNull;

import net.consentmanager.cm_sdk_android_v3.CMPManagerDelegate;
import net.consentmanager.cm_sdk_android_v3.ConsentLayerUIConfig;
import net.consentmanager.cm_sdk_android_v3.UrlConfig;
import net.consentmanager.java.JavaCMPManager;

import kotlinx.serialization.json.JsonObject;

import android.webkit.WebView;
import android.widget.FrameLayout;

public class MainActivity extends ComponentActivity implements CMPManagerDelegate {

    private JavaCMPManager cmpManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        WebView webView = new WebView(this);
        FrameLayout layout = new FrameLayout(this);
        layout.addView(webView);
        setContentView(layout);

        initializeCMPManager();
    }

    private void initializeCMPManager() {
        UrlConfig urlConfig = new UrlConfig(
                "09cb5dca91e6b",
                "delivery.consentmanager.net",
                "EN",
                "CMDemoAppJava"
        );

        ConsentLayerUIConfig webViewConfig = new ConsentLayerUIConfig(
                ConsentLayerUIConfig.Position.FULL_SCREEN,
                ConsentLayerUIConfig.BackgroundStyle.dimmed(Color.BLACK, 0.5f),
                10f,
                true,
                false
        );

        WebView webView = new WebView(this);
        cmpManager = JavaCMPManager.getInstance(this, urlConfig, webViewConfig, this);
        cmpManager.setActivity(this);
        cmpManager.setWebView(webView);

        cmpManager.checkWithServerAndOpenIfNecessary(result -> {
            if (result.isSuccess()) {
                Log.d("JavaDemoApp", "CMP Initialized successfully");
            } else {
                Log.e("JavaDemoApp", "Initialize method failed with error: " + result.exceptionOrNull());
            }
            return null;
        });
    }

    private void showCMPDemoScreen() {
        Intent intent = new Intent(this, CMPDemoActivity.class);
        startActivity(intent);
        finish();
    }

    @Override
    public void didShowConsentLayer() {
        Log.d("CMP JavaDemoApp", "Consent Layer open message received.");
    }

    @Override
    public void didCloseConsentLayer() {
        Log.d("CMP JavaDemoApp", "Consent Layer close message received.");
    }

    @Override
    public void didReceiveError(@NonNull String error) {
        Log.e("CMP JavaDemoApp", "SDK error: " + error);
    }

    @Override
    public void didReceiveConsent(@NonNull String consent, @NonNull JsonObject jsonObject) {
        Log.d("CMP JavaDemoApp", "Consent received: " + consent);
        runOnUiThread(this::showCMPDemoScreen);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (cmpManager != null) {
            cmpManager.onActivityDestroyed();
        }
    }
}

File: Kotlin/app/src/main/java/net/consentmanager/kmm/cmpsdkdemoapp/MainActivity.kt
package net.consentmanager.kmm.cmpsdkdemoapp

import android.content.res.Configuration
import net.consentmanager.kmm.cmpsdkdemoapp.BuildConfig
import android.os.Bundle
import android.util.Log
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.activity.enableEdgeToEdge
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import com.google.firebase.Firebase
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.analytics.analytics
import net.consentmanager.cm_sdk_android_v3.CMPManager
import net.consentmanager.cm_sdk_android_v3.CMPManagerDelegate
import net.consentmanager.cm_sdk_android_v3.UrlConfig

/** Intent extra to skip ConfigurationScreen (for UI tests). */
const val EXTRA_SKIP_CONFIG = "skip_config"

class MainActivity : ComponentActivity(), CMPManagerDelegate {
    private var cmpManager: CMPManager? = null
    private lateinit var analytics: FirebaseAnalytics

    private var configuration by mutableStateOf(CMPConfiguration.default)
    private var hasConfiguration by mutableStateOf(false)
    private var isConsentInitialized by mutableStateOf(false)

    private val skipConfig: Boolean
        get() = intent?.getBooleanExtra(EXTRA_SKIP_CONFIG, false) == true

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()

        analytics = Firebase.analytics

        if (skipConfig) {
            hasConfiguration = true
        }

        setContent {
            MaterialTheme {
                Surface(
                    modifier = Modifier.fillMaxSize(),
                    color = MaterialTheme.colorScheme.background
                ) {
                    when {
                        isConsentInitialized -> {
                            cmpManager?.let { CMPDemoScreen(it) }
                        }
                        hasConfiguration -> {
                            ConsentLoadingScreen(onInit = { initCMPAndOpenConsent() })
                        }
                        else -> {
                            ConfigurationScreen(
                                configuration = configuration,
                                onConfigurationChange = { configuration = it },
                                onLoadCMP = {
                                    if (configuration.isValid) {
                                        hasConfiguration = true
                                    }
                                }
                            )
                        }
                    }
                }
            }
        }
    }

    @Composable
    private fun ConsentLoadingScreen(onInit: () -> Unit) {
        LaunchedEffect(Unit) { onInit() }
        Box(
            modifier = Modifier.fillMaxSize(),
            contentAlignment = Alignment.Center
        ) {
            CircularProgressIndicator()
        }
    }

    private fun initCMPAndOpenConsent() {
        if (cmpManager != null) {
            checkAndOpenConsentLayer()
            return
        }
        if (!configuration.isValid) return
        val urlConfig = configuration.toUrlConfig()
        val webViewConfig = configuration.toConsentLayerUIConfig()
        cmpManager = CMPManager.getInstance(
            context = this,
            urlConfig = urlConfig,
            webViewConfig = webViewConfig,
            delegate = this
        )
        cmpManager!!.setActivity(this)
        checkAndOpenConsentLayer()
    }

    private fun checkAndOpenConsentLayer() {
        cmpManager?.checkAndOpen(false) { result ->
            result.onSuccess {
                runOnUiThread { isConsentInitialized = true }
            }.onFailure { error ->
                if (BuildConfig.DEBUG) {
                    Log.e("DemoApp", "Check and open consent layer failed with error: $error")
                }
            }
        }
    }

    private fun showCMPDemoScreen() {
        isConsentInitialized = true
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        if (BuildConfig.DEBUG) Log.d("CMP DemoApp", "Configuration changed")
        super.onConfigurationChanged(newConfig)
        cmpManager?.onApplicationResume()
    }

    override fun onPause() {
        super.onPause()
        cmpManager?.onApplicationPause()
    }

    override fun onDestroy() {
        super.onDestroy()
        cmpManager?.onActivityDestroyed()
    }

    override fun didShowConsentLayer() {
        if (BuildConfig.DEBUG) Log.d("CMP DemoApp", "Consent Layer open message received.")
    }

    override fun didCloseConsentLayer() {
        if (BuildConfig.DEBUG) Log.d("CMP DemoApp", "Consent Layer close message received.")
        runOnUiThread { showCMPDemoScreen() }
    }

    override fun didReceiveConsent(consent: String, jsonObject: Map<String, Any>) {
        if (BuildConfig.DEBUG) Log.d("CMP DemoApp", "Consent Layer successfully received consent message.")
        runOnUiThread { showCMPDemoScreen() }
    }

    override fun didReceiveError(error: String) {
        if (BuildConfig.DEBUG) Log.e("CMP DemoApp", "SDK error: $error")
    }
}
